### PR TITLE
add the ability to specify the decoders

### PR DESCRIPTION
**Pull request by WilsonGregory**
## Changes
- add comments for the form_block, encoder, and decoder functions
- allow the benchmark function to take the eval_model used in `new_model_metrics`, defaults to none which in turn will use a random forest classifier
- delete the `VAE_Gumbel_NInsta` class. As far as I can tell, this wasn't used anywhere. I can't tell how it is different from the other ones. For cleanliness I would like to remove old out of date code, however if this does have a purpose I can easily add it back, just let me know.
- allow specifying what decoder to use in all VAE family models
- decoder and dec_logvar can be specified individually, although likely they should always be the same

## Testing
- tested using different eval models in benchmark
- ~~tested using different decoders in MarkerMap. Strangely enough, it seems that the bernoulli decoder is outperforming the gaussian decoder in some instances. We are still investigating.~~ It appears this was just by chance, in the supervised setting the decoder and dec_logvar is not used because those are only used for the reconstruction.
image

## Doc Changes
- none

### CODE REVIEW

File: src/markermap/utils.py [hunk 1]
```diff
@@ -522,7 +560,17 @@ def __init__(self, input_size, hidden_layer_size, z_size, output_size = None, bi
...
+            decoder = make_gaussian_decoder(
+                output_size,
+                hidden_layer_size,
+                z_size,
+                bias = bias,
+                batch_norm = batch_norm,
+            )
+
+        self.decoder = decoder
+        self.dec_logvar = deepcopy(decoder)
```
[the remark this answers is not in the document]
Comment (beelze-b) [reply src/markermap/utils.py:573]: I think this might introduce a bug, or actually a new feature.

This ties the initial weights of the two decoders, whereas that is not the case in the original. Not sure if an improvement or not, but it might cause issues with reproducing results.

**Comment by beelze-b** [comment]
I think this mostly looks good. Added comments were necessary. I think we can merge after resolving those.

I am a bit concerned about gaussian decoder not being the best, but I guess that is whatever.

I don't think we need a dec_logvar if using Bernoulli decoder, but we do need to use a different supervised loss.

File: src/markermap/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (WilsonGregory) [reply src/markermap/utils.py:573]: Ah, that is a good catch. We probably want to avoid that. I guess I will just allow users to set decoder and dec_logvar separately.

**Comment by WilsonGregory** [comment]
Thanks for the feedback, I think my results here were a fluke, the decoder that I was setting to the bernoulli decoder isn't even going to be used in the supervised setting that I was using it for. Sometimes the supervised mm does badly on zeisel, and I think it may just have not done badly in the bernoulli runs just by luck. I will keep investigating this.

For your results in the other thread, I have not tried to reproduce those yet. I will try to take a look.

**Comment by beelze-b** [comment]
>  the decoder that I was setting to the bernoulli decoder isn't even going to be used in the supervised setting that I was using it for

As in switching the supervised loss from a Gaussian reconstruction loss to a cross entropy loss for the ending sigmoids (with apprpriate data pre-processing for the data)?

Even if the results were a fluke, presenting more options can be useful in the future, so I still think this pull request could be merged with a little more work.

> For your results in the other thread, I have not tried to reproduce those yet. I will try to take a look.

Yeah would be really nice to finally verify if it was just a versioning issue.

**Comment by WilsonGregory** [comment]
> As in switching the supervised loss from a Gaussian reconstruction loss to a cross entropy loss for the ending sigmoids (with apprpriate data pre-processing for the data)?

Yeah, I didn't realize that when loss_tradeoff == 0, the `self.decoder` is not used in MarkerMap, only the `self.classification_decoder` in the `decode` function. So setting `self.decoder` to the Bernoulli decoder shouldn't have any effect on the outcome. When I swap to the mixed setting or the unsupervised setting, the version with the bernoulli decoder performs poorly, as we would expect since the output is hit with the sigmoid layer at the end so the data will look a lot closer to +-1 than it is in reality.

So for the purposes of this PR, I just want to use the bernoulli decoder to confirm that it doesn't break when you specify a different decoder, rather than consider the bernoulli decoder as a research direction we should investigate (although maybe we should, I don't know).

**Comment by beelze-b** [comment]
> So for the purposes of this PR, I just want to use the bernoulli decoder to confirm that it doesn't break when you specify a different decoder, rather than consider the bernoulli decoder as a research direction we should investigate (although maybe we should, I don't know).

Makes sense to me. I think we can add options to provide a GaussianDecoder with Reconstruction Loss or Bernoulli Decoder with Cross Entropy loss as options and leave the efficacy for later if we leave the gaussian set up as default.

**Comment by beelze-b** [comment]
I noticed this got merged into main with the deepcopy of the decoders. Was that intentional?

File: src/markermap/utils.py [hunk 2]
```diff
@@ -522,7 +560,31 @@ def __init__(self, input_size, hidden_layer_size, z_size, output_size = None, bi
...
+            dec_logvar = make_gaussian_decoder(
+                output_size,
+                hidden_layer_size,
+                z_size,
+                bias = bias,
+                batch_norm = batch_norm,
+            )
+
+        self.decoder = decoder
+        self.dec_logvar = dec_logvar
```
Comment (WilsonGregory) [inline src/markermap/utils.py:587]: @beelze-b This was my solution to not use the `deepcopy`, I think this should be good right? You can pass either decoder and dec_logvar (and if you pass only 1, it will warn you that you should pass both). If they aren't passed, they are set to `make_gaussian_decoder` which should avoid the deepcopy issue. 

What I didn't do was make it so that you can specify the loss function like we were saying we would want to do if you were swapping to the bernoulli decoder. So right now this would really just help you pass different architectures for the gaussian decoder, like more hidden layers or something.

Comment (beelze-b) [reply src/markermap/utils.py:587]: This looks fine to me. Weird when I refreshed earlier I was seeing the old version. 

Sorry for the confusion.
